### PR TITLE
Revamp the Transforming Datasets user guide

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?

Revamp the Transforming Datasets user guide for Dataset GA.

## Related issue number

https://github.com/ray-project/ray/issues/23088

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

File: doc/source/data/doc_code/transforming_datasets.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+# Repartition the dataset to 5 blocks.
+ds = ds.repartition(5)
+# Dataset(num_blocks=5, num_rows=150,
+#         schema={sepal.length: double, sepal.width: double,
+#                 petal.length: double, petal.width: double, variety: string})
+
+# Find rows with sepal.length < 5.5 and petal.length > 3.5.
+def transform_batch(df: pandas.DataFrame) -> pandas.DataFrame:
+    return df[(df["sepal.length"] < 5.5) & (df["petal.length"] > 3.5)]
```
Comment (clarkzinzow) [inline doc/source/data/doc_code/transforming_datasets.py:31]: Nice!

File: doc/source/data/transforming-datasets.rst [hunk 2]
```diff
@@ -4,64 +4,85 @@
...
-    # -> [{'value': 0}, {'value': 2}, ...]
-
-The batch format can be specified using ``batch_format`` option, which defaults to "native",
-meaning pandas format for Arrow-compatible batches, and Python lists for other types. You
-can also specify explicitly "arrow" or "pandas" to force a conversion to that batch format.
-The batch size can also be chosen. If not given, the batch size will default to entire blocks.
+The Ray Datasets transformations take in datasets and produce new datasets.
+For example, *map* is a transformation that applies a user-defined function (UDF)
+on each row of input dataset and returns a new dataset as result. The Datasets
+transformations are **composable**. Operations can be further applied on result
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:10]:
```suggestion
transformations are **composable**. Operations can be further applied on the result
```

File: doc/source/data/transforming-datasets.rst [hunk 3]
```diff
@@ -4,64 +4,85 @@
...
+dataset, forming a chain of transformations to express more complex computations.
+Transformations are the core for expressing business logic in Datasets.
+
+Transformations
+---------------
+
+In general, we have two types of transformations:
+
+* One-to-one transformations: each input block will contribute to only one output
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:19]:
```suggestion
* **One-to-one transformations:** each input block will contribute to only one output
```

File: doc/source/data/transforming-datasets.rst [hunk 4]
```diff
@@ -4,64 +4,85 @@
...
+  block, such as :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`. In other
+  systems this may be called narrow transformations.
+* All-to-all transformations: input blocks can contribute to multiple output blocks,
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:22]:
```suggestion
* **All-to-all transformations:** input blocks can contribute to multiple output blocks,
```

File: doc/source/data/transforming-datasets.rst [hunk 5]
```diff
@@ -4,64 +4,85 @@
...
+     - | Repartition the dataset into N blocks, shuffling the data during repartition.
+   * - :meth:`ds.random_shuffle() <ray.data.Dataset.random_shuffle>`
+     - All-to-all
+     - | Randomly shuffle the elements of this dataset.
+   * -  :meth:`ds.sort() <ray.data.Dataset.sort>`
+     - All-to-all
+     - | Sort the dataset by a sortkey.
+   * -  :meth:`ds.groupby() <ray.data.Dataset.groupby>`
+     - All-to-all
+     - | Group the dataset by a groupkey.
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:54]: Excellent table!

File: doc/source/data/transforming-datasets.rst [hunk 6]
```diff
@@ -4,64 +4,85 @@
...
+   :start-after: __dataset_transformation_begin__
+   :end-before: __dataset_transformation_end__
+
+Compute Strategy
+----------------
+
+Datasets transformations are executed by either :ref:`Ray tasks <ray-remote-functions>`
+or :ref:`Ray actors <actor-guide>` across a Ray cluster. By default, Ray tasks are
+used (with ``compute="tasks"``). For transformations that require expensive setup,
+it's preferrable to use Ray actors, which is stateful and allows setup to be reused
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:77]:
```suggestion
it's preferrable to use Ray actors, which are stateful and allows setup to be reused
```

File: doc/source/data/transforming-datasets.rst [hunk 7]
```diff
@@ -4,64 +4,85 @@
...
-        BatchInferModel, compute=ActorPoolStrategy(10, 20),
-        batch_size=256, num_gpus=1)
-    # -> Map Progress (16 actors 4 pending): 100%|██████| 200/200 [00:07, 27.60it/s]
-
-    # Save the results.
-    ds.repartition(1).write_json("s3://bucket/inference-results")
+    Datasets also provides the convenience transformation methods :meth:`ds.map() <ray.data.Dataset.map>`,
+    :meth:`ds.flat_map() <ray.data.Dataset.flat_map>`, and :meth:`ds.filter() <ray.data.Dataset.filter>`,
+    which are not vectorized (slower than :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`), but
+    may be useful for development.
```
Comment (clarkzinzow) [inline doc/source/data/transforming-datasets.rst:61]: @ericl Do we also want to call out `ds.add_column()`?

**Review by clarkzinzow — APPROVED** [review]
LGTM, a few nits, and I wonder if we should call out `add_column()` as well. I love the table!

File: doc/source/data/transforming-datasets.rst (hunk 2 above)
Comment (jianoaix) [reply doc/source/data/transforming-datasets.rst]: Done.

File: doc/source/data/transforming-datasets.rst (hunk 3 above)
Comment (jianoaix) [reply doc/source/data/transforming-datasets.rst]: Done.

File: doc/source/data/transforming-datasets.rst (hunk 4 above)
Comment (jianoaix) [reply doc/source/data/transforming-datasets.rst]: Done.

File: doc/source/data/transforming-datasets.rst (hunk 6 above)
Comment (jianoaix) [reply doc/source/data/transforming-datasets.rst]: Done.

**Comment by clarkzinzow** [comment]
@jianoaix Excellent work on this, this is a massive improvement!

File: doc/source/data/transforming-datasets.rst (hunk 7 above)
Comment (jianoaix) [reply doc/source/data/transforming-datasets.rst]: Actually add_column is using map_batches.

**Comment by jianoaix** [comment]
Ping?

**Pull request merged**